File: 0-event-ingestion-kafka/count-consumer/0-typescript/src/counter/counter.controller.ts
import { Controller, Get } from "@nestjs/common"
import { EventPattern, Payload } from "@nestjs/microservices"
import { ClickEnvelope, CounterService } from "./counter.service"

/**
 * Consumes click events from Kafka and exposes the running counts over HTTP.
 *
 * `@EventPattern` binds the topic to the Kafka consumer; the `GET /counts`
 * route lets the e2e test and learners observe the deduplicated totals without
 * scraping logs.
 */
@Controller()
export class CounterController {
    constructor(private readonly counter: CounterService) {}

    /**
     * Handles one click record from the `ad-clicks` topic.
     *
     * @param data Decoded click envelope (Nest auto-parses the JSON value).
     */
    @EventPattern("ad-clicks")
    handle(@Payload() data: ClickEnvelope): void {
        this.counter.record(data)
    }

    /**
     * Returns the current per-ad counts and duplicate tally.
     */
    @Get("counts")
    counts(): {
        counts: Record<string, number>
        totalUnique: number
        duplicatesRejected: number
    } {
        return this.counter.snapshot()
    }
}

File: 0-event-ingestion-kafka/count-consumer/1-java/src/main/java/com/starci/counter/CounterController.java
package com.starci.counter;

import java.util.Map;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * Exposes the running counts over HTTP so the e2e test and learners can observe
 * the deduplicated totals without scraping logs.
 */
@RestController
public class CounterController {

    private final CounterService counter;

    public CounterController(CounterService counter) {
        this.counter = counter;
    }

    /** Returns the current per-ad counts and duplicate tally. */
    @GetMapping("/counts")
    public Map<String, Object> counts() {
        return counter.snapshot();
    }
}
